lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -4722,6 +4722,89 @@ FIRRTLType RefSubOp::inferReturnType(ValueRange operands,
       loc, "ref.sub op requires a RefType of vector or bundle base type");
 }
 
+//===----------------------------------------------------------------------===//
+// Optional Group Operations
+//===----------------------------------------------------------------------===//
+
+LogicalResult GroupOp::verify() {
+  auto groupName = getGroupName();
+  auto *parentOp = (*this)->getParentOp();
+
+  // Verify the correctness of the symbol reference.  Only verify that this
+  // group makes sense in its parent module or group.
+  auto nestedReferences = groupName.getNestedReferences();
+  if (nestedReferences.empty()) {
+    if (!isa<FModuleOp>(parentOp)) {
+      auto diag = emitOpError() << "has an un-nested group symbol, but does "
+                                   "not have a 'firrtl.module' op as a parent";
+      return diag.attachNote(parentOp->getLoc())
+             << "illegal parent op defined here";
+    }
+  } else {
+    auto parentGroup = dyn_cast<GroupOp>(parentOp);
+    if (!parentGroup) {
+      auto diag = emitOpError()
+                  << "has a nested group symbol, but does not have a '"
+                  << getOperationName() << "' op as a parent'";
+      return diag.attachNote(parentOp->getLoc())
+             << "illegal parent op defined here";
+    }
+    auto parentGroupName = parentGroup.getGroupName();
+    if (parentGroupName.getRootReference() != groupName.getRootReference() ||
+        parentGroupName.getNestedReferences() !=
+            groupName.getNestedReferences().drop_back()) {
+      auto diag = emitOpError() << "is nested under an illegal group";
+      return diag.attachNote(parentGroup->getLoc())
+             << "illegal parent group defined here";
+    }
+  }
+
+  // Ensure that only passive types are cpatured.
+  Block *body = getBody(0);
+  for (auto &op : *body) {
+    for (auto operand : op.getOperands()) {
+      if (operand.getParentBlock() == body)
+        continue;
+      FIRRTLBaseType baseType = dyn_cast<FIRRTLBaseType>(operand.getType());
+      if (!baseType) {
+        auto diag = emitOpError()
+                    << "captures an operand which is not a FIRRTL base type";
+        diag.attachNote(operand.getLoc()) << "operand is defined here";
+        return diag.attachNote(op.getLoc()) << "operand is used here";
+      }
+      if (!baseType.isPassive()) {
+        auto diag = emitOpError()
+                    << "captures an operand which is not a passive type";
+        diag.attachNote(operand.getLoc()) << "operand is defined here";
+        return diag.attachNote(op.getLoc()) << "operand is used here";
+      }
+    }
+    // Ensure that the group does not drive any sinks.
+    if (auto connect = dyn_cast<FConnectLike>(op)) {
+      if (connect.getDest().getParentBlock() == body)
```

Comment:
This doesn't reject connections to subelements so we need something like:
```suggestion
      if (getFieldRefFromValue(connect.getDest()).getValue().getParentBlock() == body)
```

Reply:
Good catch! I'll add this and a test.